wallet: derivehdkey RPC to get xpub at arbitrary path

Sjors opened the pull request:
Adds a `derivehdkey` RPC that returns an xpub, or optionally the xprv, at an arbitrary BIP32 path (with at least one hardened step), derived from a wallet HD key.

The main use case is coordinating a multisig setup, where each participant shares an xpub derived at a hardened path (e.g. `m/87h/0h/0h`) distinct from their default single-signature descriptors. See the (updated) `doc/multisig-tutorial.md` and (updated) functional test to see how that workflow improves.

The first commits are some helpful helpers:

- _key: add DeriveExtKey() helper_ - performs the actual derivation
- _test: move parse_hd_keypath test to bip32_tests_ - from `psbt_wallet_tests`
- _Have ParseHDKeypath handle h derivation marker_
- _util: reject out-of-range BIP32 keypath indices_ -  `ParseHDKeypath` would previously map overflowing values without `h` to hardened.
- _fuzz: check ParseHDKeypath/WriteHDKeypath round-trip_
- _rpc: ParsePathBIP32 helper_
- _refactor: add hardened derivation helper_ - `HasHardenedDerivation()`, to enforce the "at least one hardened step" rule
- _wallet: generalize GetActiveHDPubKeys helper_ - extracts code from `gethdkeys` which `derivehdkey` needs
- _wallet: add GetExtKey helper_ - reconstruct an xprv from a wallet xpub (analog of `GetKey()`); behavior-preserving prep, also simplifies `gethdkeys`.

Meat and potatoes:
- _rpc: add derivehdkey_ - the RPC itself, plus the `UnusedKey` filter on `GetHDPubKeys` that drives key selection.
- _test: use derivehdkey in M-of-N multisig demo_ - rewrites the functional multisig test to use the RPC and `<0;1>` syntax.
- _doc: use derivehdkey in multisig tutorial_ - same for the prose tutorial.

Sjors committed:
rpc: ParsePathBIP32 helper

Sjors committed:
key: add DeriveExtKey() helper

Co-authored-by: w0xlt <94266259+w0xlt@users.noreply.github.com>

Sjors committed:
Have ParseHDKeypath handle h derivation marker

Sjors committed:
test: use derivehdkey in M-of-N multisig demo

Use derivehdkey instead of extracting each participant xpub (and
derivation info) from  the listdescriptors output.

Additionally use the new <0;1> descriptor syntax.

Finally this commits adds a few debug log lines, and expand the
explanation for why we use m/44h/1h/0h.

rkrux commented:
Very nice, Concept ACK.
I will review this PR.

Sjors commented:
I pushed an extra commit 9814b3350df3ec8381a04189d66e5a20d9414ab0 to pre-empt IWYU warnings (at least in https://github.com/Sjors/bitcoin/pull/91, probably here too). That caused CI to pick up the latest master, which fails as expected. I'll push again after #29136 is updated.

Sjors commented:
Rebased after #29136 updated and squashed 9814b3350df3ec8381a04189d66e5a20d9414ab0 IWYU fixup into the first commit.

Sjors commented:
Rebased now that #29136 landed.

pseudoramdom commented:
Concept ACK

Hi @Sjors, could the implementation move out of the RPC into a `wallet::DeriveHDKey()` helper, similar to #34861 ? 
The use case I have in mind is multisig setup in gui-qml, where each cosigner needs to share a derived xpub. A single `interfaces::Wallet` method that combines `addhdkey` & `derivehdkey` would let the GUI get a shareable xpub in one call.

pseudoramdom commented:
> A single `interfaces::Wallet` method that combines `addhdkey` & `derivehdkey` would let the GUI get a shareable xpub in one call.

After chatting with @achow101, these might be separate interfaces. We'll still need the logic to be moved out of the RPC for `derivehdkey`. 
I have PR for addHDKey interface https://github.com/bitcoin/bitcoin/pull/35436

w0xlt commented on src/rpc/util.cpp:
```diff
@@ -1379,6 +1382,31 @@ std::vector<CScript> EvalDescriptorStringOrObject(const UniValue& scanobject, Fl
+std::optional<std::pair<CExtKey, KeyOriginInfo>> DeriveExtKey(const CExtKey& ext_key, const std::vector<uint32_t>& path) {
+    CExtKey descendant = ext_key;
+    KeyOriginInfo origin;
+    origin.clear(); // Prevent spurious uninitialized variable warning
```
`derivehdkey("m")` currently leaves the origin fingerprint as `[00000000]` because the fingerprint is only populated inside the child-derivation loop, which does not run for an empty path.
  
```diff
diff --git a/src/rpc/util.cpp b/src/rpc/util.cpp
index 9cae5b8ecd..1754d26cfa 100644
--- a/src/rpc/util.cpp
+++ b/src/rpc/util.cpp
@@ -1386,14 +1386,11 @@ std::optional<std::pair<CExtKey, KeyOriginInfo>> DeriveExtKey(const CExtKey& ext
     CExtKey descendant = ext_key;
     KeyOriginInfo origin;
     origin.clear(); // Prevent spurious uninitialized variable warning
+    const CKeyID id = ext_key.key.GetPubKey().GetID();
+    std::copy(id.begin(), id.begin() + sizeof(origin.fingerprint), origin.fingerprint);
     origin.path = path;
-    bool first = true;
     for (uint32_t i : path) {
         if (!descendant.Derive(descendant, i)) return std::nullopt;
-        if (first) {
-            memcpy(origin.fingerprint, descendant.vchFingerprint, 4);
-            first = false;
-        }
     }
     return std::make_pair(descendant, origin);
 }
diff --git a/src/rpc/util.h b/src/rpc/util.h
index d00920a9eb..70bde19dfe 100644
--- a/src/rpc/util.h
+++ b/src/rpc/util.h
@@ -159,7 +159,7 @@ std::vector<CScript> EvalDescriptorStringOrObject(const UniValue& scanobject, Fl
 //! Get extended key and origin info for a given path
 //! @param[in] ext_key The extended private key to derive from
 //! @param[in] path The BIP 32 path
-//! @return the resulting extended private key and origin info (blank if path is empty)
+//! @return the resulting extended private key and origin info
 std::optional<std::pair<CExtKey, KeyOriginInfo>> DeriveExtKey(const CExtKey& ext_key, const std::vector<uint32_t>& path);
 
 //! Parse BIP32 path
diff --git a/test/functional/wallet_derivehdkey.py b/test/functional/wallet_derivehdkey.py
index 118726b94b..0bedc985a8 100755
--- a/test/functional/wallet_derivehdkey.py
+++ b/test/functional/wallet_derivehdkey.py
@@ -35,6 +35,8 @@ class WalletDeriveHDKeyTest(BitcoinTestFramework):
         xpub_info = wallet.derivehdkey("m")
         assert "xprv" not in xpub_info
         xpub = xpub_info["xpub"]
+        root_fingerprint = wallet.derivehdkey("m/0")["origin"][1:9]
+        assert_equal(xpub_info["origin"], f"[{root_fingerprint}]")
 
         xpub_info = wallet.derivehdkey("m", private=True)
         xprv = xpub_info["xprv"]
```

Sjors replied:
5cedf369d976da64d022c377f9e75cfa0b5978c5: I think this lets me drop `origin.clear()` too, so I'll try that.

w0xlt commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -916,6 +917,132 @@ RPCMethod addhdkey()
+            CExtKey xprv(xpub, *key);
+
+            std::optional<std::pair<CExtKey, KeyOriginInfo>> child{DeriveExtKey(xprv, path)};
+            CHECK_NONFATAL(child);
```
Could `derivehdkey()` return `RPC_INVALID_PARAMETER` when `DeriveExtKey()` returns null for a user-provided path, such as one exceeding BIP32’s maximum depth?

That would avoid surfacing invalid input through `CHECK_NONFATAL()`, which looks more like an internal bug path.

```diff
diff --git a/src/wallet/rpc/wallet.cpp b/src/wallet/rpc/wallet.cpp
index f086ead493..87a860dbf8 100644
--- a/src/wallet/rpc/wallet.cpp
+++ b/src/wallet/rpc/wallet.cpp
@@ -1027,7 +1027,9 @@ RPCMethod derivehdkey()
             CExtKey xprv(xpub, *key);
 
             std::optional<std::pair<CExtKey, KeyOriginInfo>> child{DeriveExtKey(xprv, path)};
-            CHECK_NONFATAL(child);
+            if (!child) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Unable to derive HD key at the requested path");
+            }
 
             UniValue res{UniValue::VOBJ};
 
diff --git a/test/functional/wallet_derivehdkey.py b/test/functional/wallet_derivehdkey.py
index 118726b94b..13220a5aba 100755
--- a/test/functional/wallet_derivehdkey.py
+++ b/test/functional/wallet_derivehdkey.py
@@ -35,6 +35,13 @@ class WalletDeriveHDKeyTest(BitcoinTestFramework):
         xpub_info = wallet.derivehdkey("m")
         assert "xprv" not in xpub_info
         xpub = xpub_info["xpub"]
+        too_deep_path = "m/" + "/".join(["0"] * 256)
+        assert_raises_rpc_error(
+            -8,
+            "Unable to derive HD key at the requested path",
+            wallet.derivehdkey,
+            too_deep_path,
+        )
 
         xpub_info = wallet.derivehdkey("m", private=True)
         xprv = xpub_info["xprv"]
```

w0xlt reviewed (commented):
A few review comments:

Sjors commented:
I'll mark this draft pending https://github.com/bitcoin/bitcoin/pull/35436.

Sjors commented:
I tried basing this on #35436, but the current version has too many issues. So instead I just rebased on master and applied @w0xlt's suggestions. Ready for review again.

(I do still think it's good to move implementation code out of RPC, but this can be done later)

Sjors commented on src/util/bip32.cpp:
```diff
@@ -17,7 +17,7 @@ bool ParseHDKeypath(const std::string& keypath_str, std::vector<uint32_t>& keypa
     std::string item;
     bool first = true;
-    while (std::getline(ss, item, '/')) {
+    while (std::getline(ss, item, '/') || std::getline(ss, item, 'h')) {
```
6e0efac98d3325e9d699ad1a92bd2a5324fc41bd: I'm adding test coverage in the next push.

Sjors commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -918,6 +919,134 @@ RPCMethod addhdkey()
+                        w_desc.descriptor->GetPubKeys(desc_pubkeys, desc_xpubs);
+                        for (const CExtPubKey& xpub : desc_xpubs) {
+                            std::string desc_str;
+                            bool ok = desc_spkm->GetDescriptorString(desc_str, false);
```
In dc57236474f13ce3e62081bbb763dc90be1c9a13 _rpc: add derivehdkey_: I copy-pasted this check from `gethdkeys`, but it's not needed here. Will drop in the next push.

Sjors commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -918,6 +919,134 @@ RPCMethod addhdkey()
+                    }
+
+                    if (wallet_xpubs.size() == 0) {
+                        throw JSONRPCError(RPC_INVALID_PARAMETER, "No active or unused(KEY) descriptor found");
```
In dc57236474f13ce3e62081bbb763dc90be1c9a13 _rpc: add derivehdkey_: I'm switching this to `RPC_INVALID_ADDRESS_OR_KEY`.

Sjors committed:
doc: use derivehdkey in multisig tutorial

Use derivehdkey instead of extracting each participant xpub
from  the listdescriptors output.

Additionally use the new <0;1> descriptor syntax.

Sjors commented:
The above inline comments were from _before_ I pushed and announce some small changes. I also slightly improved the tutorial.

rkrux commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -918,6 +919,133 @@ RPCMethod addhdkey()
+
+            if (!wallet->IsWalletFlagSet(WALLET_FLAG_DESCRIPTORS)) {
+                throw JSONRPCError(RPC_WALLET_ERROR, "derivehdkey is not available for non-descriptor wallets");
+            }
```
In b3ec19f720f9ddfd77fe831aadb9c9f8c3ab1d76 "rpc: add derivehdkey"

```diff
diff --git a/src/wallet/rpc/wallet.cpp b/src/wallet/rpc/wallet.cpp
index dc2fcdb207..c299a91f16 100644
--- a/src/wallet/rpc/wallet.cpp
+++ b/src/wallet/rpc/wallet.cpp
@@ -949,10 +949,6 @@ RPCMethod derivehdkey()
             const std::shared_ptr<const CWallet> wallet = GetWalletForJSONRPCRequest(request);
             if (!wallet) return UniValue::VNULL;
 
-            if (!wallet->IsWalletFlagSet(WALLET_FLAG_DESCRIPTORS)) {
-                throw JSONRPCError(RPC_WALLET_ERROR, "derivehdkey is not available for non-descriptor wallets");
-            }
-
             if (wallet->IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS)) {
                 // Watch-only wallets can't contain unused(KEY) descriptors
                 throw JSONRPCError(RPC_WALLET_ERROR, "derivehdkey is not available for watch-only wallets");
```

I think this check was relevant when the legacy wallets could be loaded, but it's not possible now and we should avoid this check in the wallet code now.

rkrux commented on doc/release-notes-32784.md:
```diff
@@ -0,0 +1,5 @@
+Wallet
+------
+
+- A new `derivehdkey` RPC is available to obtain an xpub or xpriv for any given BIP32 path.
```
In b3ec19f "rpc: add derivehdkey"

s/xpriv/xprv

rkrux commented on doc/release-notes-32784.md:
```diff
@@ -0,0 +1,5 @@
+------
+
+- A new `derivehdkey` RPC is available to obtain an xpub or xpriv for any given BIP32 path.
+  The hd key can then be imported as e.g. part of a multisig descriptor. (#22341)
```
In b3ec19f "rpc: add derivehdkey"

I feel this use case of making the core wallet more flexible in a multi-sig setup can be mentioned in some detail here.

Sjors replied:
I expanded the note and also referenced the tutorial.

rkrux commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -918,6 +919,133 @@ RPCMethod addhdkey()
+
+            // Standard derivation paths from the HD seed contain at least one
+            // hardened step, so just always unlock the wallet.
+            EnsureWalletIsUnlocked(*wallet);
```
In b3ec19f "rpc: add derivehdkey"

The non-standard paths can contain zero hardened steps. I believe it's trivial to check for hardened-ness in the derivation path so that we can unlock the wallet conditionally? In case of zero hardened steps, the private keys need not be brought into the picture.

Sjors replied:
Do we know a single use case for a path that's unhardened from the seed? It just seems unsafe and not worth supporting.

rkrux replied:
> Do we know a single use case for a path that's unhardened from the seed?

I'm not aware.

>  It just seems unsafe and not worth supporting.

Nice, not supporting gels well with the other suggestion https://github.com/bitcoin/bitcoin/pull/32784#discussion_r3466770998 where we will mandate hardened steps in the path.

rkrux commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -918,6 +919,133 @@ RPCMethod addhdkey()
+            {"path", RPCArg::Type::STR, RPCArg::Optional::NO, "BIP 32 derivation path."},
+            {"options", RPCArg::Type::OBJ_NAMED_PARAMS, RPCArg::Optional::OMITTED, "", {
+                {"private", RPCArg::Type::BOOL, RPCArg::Default{false}, "Show private key"},
+                {"hdkey", RPCArg::Type::STR, RPCArg::DefaultHint{"Either the HD key used by all other active descriptors, or an unused(KEY) descriptor"}, "The HD key that the wallet knows the private key of, listed using 'gethdkeys', to use for derivation"},
```
In b3ec19f "rpc: add derivehdkey"

Regarding the option of displaying the derived/child private key: I'm concerned about the accidental leakage of the parent/root private key in case a non-hardened derivation path is used. The risk is non-obvious, that's why `listdescriptors private=True` also doesn't display the child private key, but it instead shows the master private key with the derivation path (which as well should not be an option here for this RPC I believe).

https://github.com/bitcoin/bitcoin/blob/d84fc352cbc1363df5cd6024a22e73fc63283e4f/src/wallet/scriptpubkeyman.cpp#L1560-L1565 

If the sole motivation of this PR is to make the core wallet flexible in multi-sig setups where only the xpubs need to be shared with other participants, then let's not provide the private option altogether?

> Either the HD key used by all other active descriptors, or an unused(KEY) descriptor"

If only the unused descriptor was considered, then this risk could have been smaller but given that the HD key used by all other active descriptors is also considered, then I feel the risk is quite high.

rkrux replied:
If there is any other reason to display the child private key, then I really feel we should mandate hardened-ness in the derivation path.

Sjors replied:
> then let's not provide the private option altogether?

You need the xprv when importing a descriptor where Core is one of the signers. Unfortunately we don't recognize an imported xpub as our own when it derives from our hd key. That requires something like https://github.com/Sjors/bitcoin/pull/91/changes/9f57f8705b71e06c868a59d28cb2998976df6162.

> we should mandate hardened-ness in the derivation path

That's fine by me.

Sjors replied:
I introduced a `HasHardenedDerivation` helper (extracted from descriptor code) and use it to enforce that there's at least one hardened derivation.

rkrux replied:
> You need the xprv when importing a descriptor where Core is one of the signers. Unfortunately we don't recognize an imported xpub as our own when it derives from our hd key. That requires something like https://github.com/Sjors/bitcoin/commit/9f57f8705b71e06c868a59d28cb2998976df6162.

Can we add that commit in this PR or raise a separate PR containing that commit (making this one dependent on that one)? I'd prefer to avoid returning any kind of private key (master or derived) just for setting up a multi-sig wallet where the core wallet is one of the signers - I'm not a proponent of making the user handle their private keys manually. It seems quite reasonable for the wallet to detect any derived xpub that's being imported.

Sjors replied:
I don't think we should make this PR more complicated. I'll look into making it a standalone PR, but would prefer to not make this one dependent on it.

We generally allow exporting private keys. Note that the multisig tutorial only uses the xpub, so we're not encouraging it. I'd like to eventually add an example of using a Bitcoin Core hot key, but that requires additional changes (e.g. #33112).

Sjors replied:
Also it doesn't seem like 9f57f8705b71e06c868a59d28cb2998976df6162 is enough, we probably also need 0a11232c137ceb6c5b9f4a894baff8d1d1042b13, 90ae3563559e1516807c2d46e74432c378629fb0 and 7bde38b0fde9f91821e7b9cb011761c17d96e58c. All of which are heavily vibed coded bits of #91, so they'll need cleanup in any case. You're welcome to give it a try by the way.

rkrux replied:
If multiple commits are required, then I guess it's fine to not increase the diff of this PR.

rkrux commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -918,6 +919,133 @@ RPCMethod addhdkey()
+            }
+
+            // If hdkey was not specified, try to look it up. First consider the
+            // active descriptors. Otherwise look for an unused(KEY) descriptor.
```
In b3ec19f "rpc: add derivehdkey"

Can we switch the order here? I'd prefer to use the unused descriptors (if present) instead of combining it with the main hd key that powers all the active descriptors. This might encourage the usage of the addhdkey RPC as well while providing a clean separation.

Sjors replied:
Done, that makes much more sense.

rkrux commented on src/rpc/util.cpp:
```diff
@@ -1374,6 +1376,18 @@ std::vector<CScript> EvalDescriptorStringOrObject(const UniValue& scanobject, Fl
     return ret;
 }
 
+std::optional<std::pair<CExtKey, KeyOriginInfo>> DeriveExtKey(const CExtKey& ext_key, const std::vector<uint32_t>& path) {
```
In 43c7770b63b46c06d27a09616f13c6178034c8cc "rpc: add DeriveExtKey() helper"

This seems like a method that could prove to be generic in the future. Can we put this in src/key.cpp instead so that its presence is highlighted - it doesn't need to be a part of the CExtKey class. I'm unable to think why its presence should be restricted within src/rpc/ only.

Sjors replied:
`src/key.cpp` is part of the kernel, which we shouldn't add things to that aren't needed for validation. So we'd have to add `src/common/key.{h,cpp}`. I think that's better done as a refactor if someone wants to reuse this outside of RPC.

rkrux replied:
> src/key.cpp is part of the kernel, which we shouldn't add things to that aren't needed for validation

I checked in src/kernel/CMakeLists.txt that only src/pubkey.cpp is a part of it and not src/key.cpp. Which kind of makes sense(?) because validation requires the Verify method that belongs in CPubKey, and not in CKey/CExtKey.

https://github.com/bitcoin/bitcoin/blob/d84fc352cbc1363df5cd6024a22e73fc63283e4f/src/kernel/CMakeLists.txt#L47

Sjors replied:
Good point, let me see if I can move it... (just missed this in the last push)

Sjors replied:
Done in 0973920308c300d2584ff2ddb4c06e83c8d09fc6, also added a direct test.

rkrux reviewed (commented):
Partial code review ff0ade21fc67e9a7ec84f30ea2bea81d5946a2e5

> The above inline comments were from before I pushed and announce some small changes. 

Can these comments be resolved now? They came across confusing to me at the first glance.

Sjors committed:
refactor: add hardened derivation helper

Sjors commented:
Ok, this should address the last review round: https://github.com/bitcoin/bitcoin/pull/32784#pullrequestreview-4561263861

pseudoramdom commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -918,6 +921,129 @@ RPCMethod addhdkey()
+                }
+            }
+
+            std::optional<CKey> key{wallet->GetKey(xpub.pubkey.GetID())};
```
In 63450ca4b218aede226e00df5df5a0267f9f76b6 rpc: add derivehdkey

IIUC this is checking if the wallet has the private key for `xpub.pubkey`.
Should we verify that the supplied xpub exactly matches one of the wallet’s known descriptor xpubs before using it.

Sjors replied:
`CExtPubKey xpub` comes from the `hdkey` argument of the RPC call. It's the xpub for the master xprv. It's not expected to appear in regular wallet descriptors, because those use xpubs at a deeper derivation path (usually account level). The only exception is the new `unused()` descriptor.

However it does make sense to require that the xpub matches either an `unused()` descriptor or an xpub corresponding to an hd key. I added that check plus a test (63450ca4b218aede226e00df5df5a0267f9f76b6 -> ...). I first extracted two news helpers: `GetExtKey` in b40725823822275139315631e9f9f914eb0c87c0 and `GetHDPubKeys` in f488a7711074001523b7ab6f97fc7e173c1f059f.

Sjors committed:
wallet: generalize GetActiveHDPubKeys helper

GetHDPubKeys() centralizes the descriptor xpub lookup used by gethdkeys and
createwalletdescriptor, and by the derivehdkey RPC added in a later commit.
The HDKeyFilter argument serves gethdkeys' active_only mode (Active vs All)
and createwalletdescriptor's active descriptor selection.

No behavior change, except the dynamic_cast now uses Assert() instead of
gethdkeys' CHECK_NONFATAL, since it is not a recoverable input.

Sjors committed:
rpc: add derivehdkey

Add an UnusedKey filter to GetHDPubKeys() so the new RPC can prefer
unused(KEY) descriptors before falling back to active descriptors.

Co-authored-by: w0xlt <94266259+w0xlt@users.noreply.github.com>

Sjors committed:
wallet: add GetExtKey helper

Reconstruct a descriptor's extended private key from its xpub by looking
up the corresponding private key. This is the extended-key analog of
GetKey() and is used by the derivehdkey RPC in the following commit.

Sjors commented:
Addressed @pseudoramdom's feedback https://github.com/bitcoin/bitcoin/pull/32784#discussion_r3478255839. This adds a safety check _and_ gets rid of code duplication in the process. I also expanded the PR description.

Some commits could be a standalone refactor pull request if the review burden is too much, but they all towards the goal of adding `derivehdkey`.

w0xlt commented on src/rpc/util.cpp:
```diff
@@ -1374,6 +1375,15 @@ std::vector<CScript> EvalDescriptorStringOrObject(const UniValue& scanobject, Fl
+    std::vector<uint32_t> out;
+    if (!ParseHDKeypath(path, out)) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid BIP32 keypath");
+    }
```
`ParsePathBIP32()` currently accepts `m/2147483648`, which is `0x80000000`. Since BIP32 uses that high bit to mark hardened derivation, the RPC ends up treating this as `m/0h` instead of rejecting it.

Suggestion:


<details>
<summary>diff</summary>

```diff
diff --git a/src/rpc/util.cpp b/src/rpc/util.cpp
index ab27a84a67..d1770f0bdf 100644
--- a/src/rpc/util.cpp
+++ b/src/rpc/util.cpp
@@ -29,6 +29,7 @@
 
 #include <algorithm>
 #include <iterator>
+#include <sstream>
 #include <string_view>
 #include <tuple>
 #include <utility>
@@ -1381,6 +1382,25 @@ std::vector<uint32_t> ParsePathBIP32(const std::string& path)
     if (!ParseHDKeypath(path, out)) {
         throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid BIP32 keypath");
     }
+
+    // BIP32 child indices must fit in 31 bits; the high bit is reserved for
+    // the hardened derivation marker.
+    std::stringstream ss(path);
+    std::string item;
+    bool first{true};
+    while (std::getline(ss, item, '/')) {
+        if (first && item == "m") {
+            first = false;
+            continue;
+        }
+        if (item.ends_with('\'') || item.ends_with('h')) item.pop_back();
+
+        const auto number{ToIntegral<uint32_t>(item)};
+        if (!number || *number > 0x7fffffffU) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid BIP32 keypath");
+        }
+        first = false;
+    }
     return out;
 }
 
diff --git a/test/functional/wallet_derivehdkey.py b/test/functional/wallet_derivehdkey.py
index 5ed8cc0eb0..32287c9464 100755
--- a/test/functional/wallet_derivehdkey.py
+++ b/test/functional/wallet_derivehdkey.py
@@ -44,6 +44,13 @@ class WalletDeriveHDKeyTest(BitcoinTestFramework):
             wallet.derivehdkey,
             "m/87/0/0",
         )
+        for path in ["m/2147483648", "m/2147483648h", "m/2147483648'"]:
+            assert_raises_rpc_error(
+                -8,
+                "Invalid BIP32 keypath",
+                wallet.derivehdkey,
+                path,
+            )
         assert_raises_rpc_error(
             -5,
             "No active or unused(KEY) descriptor found",
```
</details>

Sjors replied:
Good catch! Fixed in f086c97ab56d04f48297cb0ec1029ff61f2ff772 (in `bip32.cpp`)

Sjors committed:
test: move parse_hd_keypath test to bip32_tests

ParseHDKeypath() lives in util/bip32, so its unit test belongs in
bip32_tests rather than psbt_wallet_tests. Pure move, no changes to the
test itself; subsequent commits extend it in its new home.

Sjors committed:
util: reject out-of-range BIP32 keypath indices

ParseHDKeypath() parsed each path element with ToIntegral<uint32_t>, so
a bare decimal >= 2^31 (e.g. "m/2147483648" == 0x80000000) was silently
treated as "m/0h".

This commit rejects such overflow instead.

Sjors commented:
Added two commits:
- 58f484d2dc8126b4b73f3418dbda8c8181cca945 _test: move parse_hd_keypath test to bip32_tests_ : since this PR makes a bunch of changes to it, might as well move it to a more appropriate file
  - `ParseHDKeypath` used to live in `wallet/rpcwallet.cpp` in #13557 / #13712
- 27a50f76ca571ff42094a203ef69f6a3887cb743 _util: reject out-of-range BIP32 keypath indices_ : fixes https://github.com/bitcoin/bitcoin/pull/32784#discussion_r3502575050

Sjors committed:
fuzz: check ParseHDKeypath/WriteHDKeypath round-trip

Sjors commented:
Rebased after #32489. I moved the new `fuzz: check ParseHDKeypath/WriteHDKeypath round-trip` closer to the two `ParseHDKeypath` commits in the stack.

pseudoramdom commented on src/key.cpp:
```diff
@@ -365,6 +367,19 @@ bool CExtKey::Derive(CExtKey &out, unsigned int _nChild) const {
+    CExtKey descendant = ext_key;
+    KeyOriginInfo origin;
+    const CKeyID id = ext_key.key.GetPubKey().GetID();
+    std::copy(id.begin(), id.begin() + sizeof(origin.fingerprint), origin.fingerprint);
```
In `key: add DeriveExtKey() helper` fdfb6cae47b8a118a9620ac7764fc45aff76239a

Since #35606 is merged and changes fingerprints to the fixed-size `KeyFingerprint` type, this can become
```cpp
origin.fingerprint = ext_key.id_key_fingerprint();
```

pseudoramdom commented on src/test/bip32_tests.cpp:
```diff
@@ -184,6 +185,35 @@ BOOST_AUTO_TEST_CASE(bip32_test5) {
+
+    KeyOriginInfo expected_origin;
+    const CKeyID id{master.key.GetPubKey().GetID()};
+    std::copy(id.begin(), id.begin() + sizeof(expected_origin.fingerprint), expected_origin.fingerprint);
```
Ditto. We can use 

```cpp
expected_origin.fingerprint = master.id_key_fingerprint();
```

pseudoramdom commented on src/test/bip32_tests.cpp:
```diff
@@ -11,6 +11,7 @@
 #include <test/util/setup_common.h>
 #include <util/strencodings.h>
 
+#include <algorithm>
```
This can be dropped if we use the `id_key_fingerprint()` suggestion

pseudoramdom commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -954,6 +957,124 @@ static RPCMethod exportwatchonlywallet()
+
+            UniValue res{UniValue::VOBJ};
+
+            std::string fingerprint{HexStr(std::span<unsigned char>(child->second.fingerprint, child->second.fingerprint + 4))};
```
In `rpc: add derivehdkey` (97e1757ab8632a08e48050510658b7dd5cca6657)

With the fingerprint changes from #35606 , we can simplify this to 

```cpp
const std::string fingerprint{HexStr(child->second.fingerprint)};
```

pseudoramdom reviewed (commented):
Since #35606 has now merged, we can simplify the fingerprint handling using the new type-safe helpers.

Sjors commented:
Rebased and switched to https://github.com/bitcoin/bitcoin/pull/35606's `KeyFingerprint`, as suggested by @pseudoramdom.

Sjors commented:
I think I'll just wait for #35896 to land (tested locally that it helps).

w0xlt commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -689,33 +695,19 @@ RPCMethod gethdkeys()
-                    if (std::optional<CKey> key = priv ? desc_spkm->GetKey(xpub.pubkey.GetID()) : std::nullopt) {
-                        wallet_xprvs[xpub] = CExtKey(xpub, *key);
+                    wallet_xpubs[xpub].emplace(desc_str, wallet->IsActiveScriptPubKeyMan(*desc_spkm), desc_spkm->HasPrivKey(xpub.pubkey.GetID()));
+                }
```
Can we look up the private key directly from each associated desc_spkm inside the existing loop in  `RPCMethod gethdkeys()`?

```diff
diff --git a/src/wallet/rpc/wallet.cpp b/src/wallet/rpc/wallet.cpp
index cfba900d22..aa9dae8eb7 100644
--- a/src/wallet/rpc/wallet.cpp
+++ b/src/wallet/rpc/wallet.cpp
@@ -704,10 +704,8 @@ RPCMethod gethdkeys()
                     bool ok = desc_spkm->GetDescriptorString(desc_str, /*priv=*/false);
                     CHECK_NONFATAL(ok);
                     wallet_xpubs[xpub].emplace(desc_str, wallet->IsActiveScriptPubKeyMan(*desc_spkm), desc_spkm->HasPrivKey(xpub.pubkey.GetID()));
-                }
-                if (priv) {
-                    if (std::optional<CExtKey> xprv = wallet->GetExtKey(xpub)) {
-                        wallet_xprvs[xpub] = *xprv;
+                    if (std::optional<CKey> key = priv ? desc_spkm->GetKey(xpub.pubkey.GetID()) : std::nullopt) {
+                        wallet_xprvs[xpub] = CExtKey{xpub, *key};
                     }
                 }
             }
```

Sjors replied:
I dropped the `rpc/wallet.cpp‎` change in 99bea69af7040352064440466d49eec7132e1e1c.

w0xlt commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -965,6 +957,124 @@ static RPCMethod exportwatchonlywallet()
+            },
+        },
+        RPCExamples{
+            HelpExampleCli("derivehdkey", "m/87h/0h/0h") + HelpExampleRpc("derivehdkey", "")
```
The parameter is required.

```diff
diff --git a/src/wallet/rpc/wallet.cpp b/src/wallet/rpc/wallet.cpp
index cfba900d22..33450695e1 100644
--- a/src/wallet/rpc/wallet.cpp
+++ b/src/wallet/rpc/wallet.cpp
@@ -979,7 +979,7 @@ RPCMethod derivehdkey()
             },
         },
         RPCExamples{
-            HelpExampleCli("derivehdkey", "m/87h/0h/0h") + HelpExampleRpc("derivehdkey", "")
+            HelpExampleCli("derivehdkey", "m/87h/0h/0h") + HelpExampleRpc("derivehdkey", "\"m/87h/0h/0h\"")
             + HelpExampleCliNamed("derivehdkey", {{"path", "m/87h/0h/0h"}, {"private", "true"}})
             + HelpExampleRpcNamed("derivehdkey", {{"path", "m/87h/0h/0h"}, {"private", "true"}})
         },
```

Sjors replied:
Done

Sjors commented:
Rebased after #35896, which should get rid of ci failures for `test: move parse_hd_keypath test to bip32_tests`. Took both @w0xlt's suggestions.

w0xlt commented:
Code looks good. One question following up on the earlier parent-key leakage discussion https://github.com/bitcoin/bitcoin/pull/32784#discussion_r3466765466:

`HasHardenedDerivation()` accepts `m/44h/1h/0h/0` because the path contains hardened steps. With `private=true`, the returned child xprv and public account xpub allow recovery of the account xprv.

Should private output require the final step to be hardened?

<details>
<summary>possible diff if so.</summary>

```diff
diff --git a/src/wallet/rpc/wallet.cpp b/src/wallet/rpc/wallet.cpp
index 054ab6b763..a98471c382 100644
--- a/src/wallet/rpc/wallet.cpp
+++ b/src/wallet/rpc/wallet.cpp
@@ -965,7 +965,7 @@ RPCMethod derivehdkey()
         {
             {"path", RPCArg::Type::STR, RPCArg::Optional::NO, "BIP 32 derivation path with at least one hardened step."},
             {"options", RPCArg::Type::OBJ_NAMED_PARAMS, RPCArg::Optional::OMITTED, "", {
-                {"private", RPCArg::Type::BOOL, RPCArg::Default{false}, "Show private key"},
+                {"private", RPCArg::Type::BOOL, RPCArg::Default{false}, "Show private key. The final derivation step must be hardened"},
                 {"hdkey", RPCArg::Type::STR, RPCArg::DefaultHint{"Either the HD key of an unused(KEY) descriptor, or any other active descriptor."}, "The HD key that the wallet knows the private key of, listed using 'gethdkeys', to use for derivation"},
             }},
         },
@@ -998,6 +998,9 @@ RPCMethod derivehdkey()
             if (!HasHardenedDerivation(path)) {
                 throw JSONRPCError(RPC_INVALID_PARAMETER, "Derivation path requires at least one hardened step");
             }
+            if (priv && !(path.back() >> 31)) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Private key output requires a hardened final derivation step");
+            }
 
             LOCK(wallet->cs_wallet);
 
diff --git a/test/functional/wallet_derivehdkey.py b/test/functional/wallet_derivehdkey.py
index 8905c55e68..6c741233bb 100755
--- a/test/functional/wallet_derivehdkey.py
+++ b/test/functional/wallet_derivehdkey.py
@@ -73,8 +73,8 @@ class WalletDeriveHDKeyTest(BitcoinTestFramework):
             too_deep_path,
         )
 
-        xpub_info = wallet.derivehdkey("m/87h/0h/0h/0")
-        xpub_priv_info = wallet.derivehdkey("m/87h/0h/0h/0", private=True)
+        xpub_info = wallet.derivehdkey("m/87h/0h/0h/0h")
+        xpub_priv_info = wallet.derivehdkey("m/87h/0h/0h/0h", private=True)
         xprv = xpub_priv_info["xprv"]
         assert_equal(xpub_priv_info["xpub"], xpub_info["xpub"])
 
@@ -99,11 +99,11 @@ class WalletDeriveHDKeyTest(BitcoinTestFramework):
             -13,
             "Error: Please enter the wallet passphrase with walletpassphrase first",
             wallet.derivehdkey,
-            "m/87h/0h/0h/0",
+            "m/87h/0h/0h/0h",
             private=True,
         )
         with WalletUnlock(wallet, "pass"):
-            xpub_info = wallet.derivehdkey("m/87h/0h/0h/0", private=True)
+            xpub_info = wallet.derivehdkey("m/87h/0h/0h/0h", private=True)
             # Unused(KEY) is preferred over active descriptors and is not
             # rotated on encryption.
             assert_equal(xpub_info["xprv"], prev_xprv)
@@ -140,6 +140,18 @@ class WalletDeriveHDKeyTest(BitcoinTestFramework):
         active_xpub = wallet.gethdkeys(active_only=True)[0]["xpub"]
         assert_equal(wallet.derivehdkey("m/44h/1h/0h", hdkey=active_xpub), xpub_info)
 
+        # An xprv at an unhardened child, together with its exposed parent
+        # xpub, would reveal the parent xprv. Public output remains allowed.
+        unhardened_path = "m/44h/1h/0h/0"
+        assert "xprv" not in wallet.derivehdkey(unhardened_path)
+        assert_raises_rpc_error(
+            -8,
+            "Private key output requires a hardened final derivation step",
+            wallet.derivehdkey,
+            unhardened_path,
+            private=True,
+        )
+
         # Get the activate wpkh() receive descriptor
         desc = list(filter(lambda d:
                            d["active"] and not d["internal"] and d["desc"][0:3] == "pkh",
```
</details>

Sjors commented:
> Should private output require the final step to be hardened?

I think anyone who obtains xprv's should be careful, not sure if adding restrictions makes sense.

pseudoramdom commented:
This is in a good state IMO. 
code review ACK c3945bfd2bf87ae6fe7be1c0cee58b21c4269aec

achow101 commented:
ACK c3945bfd2bf87ae6fe7be1c0cee58b21c4269aec

davidgumberg commented:
This needs a release note!

Sjors commented:
@davidgumberg it's in `doc/release-notes-32784.md`

optout21 commented:
Post-merge comment: the merge conflict on #35170 showed that it has overlap with this PR. I have updated #35170 accordingly. Reviews are welcome on it!

davidgumberg commented:
> @davidgumberg it's in `doc/release-notes-32784.md`

Oops, I looked at the commit list. My bad
